omniscidb/ArrowStorage/ArrowStorage.cpp

```diff
@@ -330,15 +341,156 @@ TableFragmentsInfo ArrowStorage::getEmptyTableMetadata(int table_id) const {
   return res;
 }
 
-const DictDescriptor* ArrowStorage::getDictMetadata(int dict_id, bool /*load_dict*/) {
+const DictDescriptor* ArrowStorage::getDictMetadata(int dict_id, bool load_dict) {
   mapd_shared_lock<mapd_shared_mutex> dict_lock(dict_mutex_);
   CHECK_EQ(getSchemaId(dict_id), schema_id_);
   if (dicts_.count(dict_id)) {
-    return dicts_.at(dict_id).get();
+    auto dict = dicts_.at(dict_id).get();
+    CHECK(dict);
+    if (!dict->is_materialized && load_dict) {
+      std::lock_guard<std::mutex> materialization_lock(dict->mutex);
+      // after we get the lock, check the atomic again
+      if (!dict->is_materialized) {
+        materializeDictionary(dict);
+      }
+    }
+    return dict->dict_descriptor.get();
   }
   return nullptr;
 }
 
+void ArrowStorage::materializeDictionary(DictionaryData* dict) {
+  CHECK(dict);
+  CHECK(!dict->table_ids.empty());
+
+  const auto dict_desc = dict->dict_descriptor.get();
+  CHECK(dict_desc);
+  auto* string_dict = dict_desc->stringDict.get();
+  CHECK(string_dict);
+  const int elem_size = dict_desc->dictNBits / 8;
+
+  for (const auto table_id : dict->table_ids) {
+    auto& table = *tables_.at(table_id);
```

Comment:
I will try -- the locking here is a bit different because we (1) already have a lock on the `dict_mutex_` in the caller and (2) we usually are accessing this method in some very different context from import -- e.g. from column fetcher, or from `setupCaching` in RelAlgExecutor. I tried to make the sequence match `appendArrowTable` initially, but that didn't work because the calling paths are different. Do we have a lock that prevents tables_ from being changed when a query is executing?

Reply:
`data_mutex_` is used only to access `tables_` map. Once you have a table pointer, you can lock the table and release the outer lock. We don't mind if someone is modifying `tables_` while some query is executing. So, someone might create a new table and that's why you need a shared lock here. Locked `dict_mutex_` actually prevents new tables creation, but that's mostly because we lock it just in case not caring if created table actually has dictionaries. To keep it simple, I'd say we should always lock `data_mutex_` to access `tables_` structure.

Reply:
Makes sense. Done.